### DISCUSSION TITLE
How are EMA's calculated on charts up to the previous candle.

### DISCUSSION THREAD

**Discussion by CyberNomadDev** [DaveSkender/Stock.Indicators#528] [Q&A] [answered]
Hi I'm to get my head round EMA's, I cant figure out how these are applied to the previous candle on charts showing live data.
I have Historic OHLCV stored in a SQL table and I add the previous candle OHLCV  to this table each minute. How are EMAs calculated on the previous candle if they are based on historic data? I've added a image of a chart that has a EMA12 and EMA26 overlaid on it to try and show what I'm trying say. Is the close price of the previous candle added to the EMA?
![Chart](https://user-images.githubusercontent.com/88987565/129492757-f36d942a-5118-434d-aa40-b498ac8e888a.png)

**Comment by DaveSkender** [comment, marked as answer]
If you examine the [formula for EMA](https://www.investopedia.com/terms/e/ema.asp) you'll see this concept of historical "smoothing".  In a nutshell:

```
EMAtoday = EMAyesterday + SmoothingMultiplier × (CloseToday - EMAyesterday)

where

SmoothingMultiplier = 2 / (lookbackPeriods + 1)

EMA(12) would have SM = 1/(12+1) = 0.091
```
If you read it carefully, EMA value is mostly derived from historical EMAs.  

The library is quite fast, so I'd suggest recalculating the whole results series instead of trying to do incremental EMA values if that was your goal.  See #396 for more discussion on incremental calculations.  I'll be looking at the streaming and incremental use cases in v2 of the library, where we'll add a method like `GetEmaIncrement(..)`.  See #218 for our feature roadmap.

As a side note, this smoothing also presents a convergence issue worth noting.  Since EMA is initialized with a Simple Moving Average, which is essentially a guess, there is also an initial warmup window where there is convergence to a precision value.  For that reason, it is good to also use the `.RemoveWarmupPeriods` extension to remove those initial values.  See [documentation](https://daveskender.github.io/Stock.Indicators/indicators/Ema/#content) for more information.

**Reply by DaveSkender** [reply]
If your question was specifically about the last value in charting platforms, I suspect it is continuously repainted until the bar time elapses.  For example, if you're using hourlong bars and your chart is refreshing live every minute, they would discard the current hour EMA value and recalculate it using the most recent tick price.

**Reply by CyberNomadDev** [reply]
Thanks for explaining the EMA's Dave, I guess what I'm asking is how are the charting platforms able to calculate EMA12 and EMA26 on corresponding current candle, like you would only be able to look back 2 places. not 12 or 26.

**Reply by DaveSkender** [reply]
Based on the formula, to do an incremental calculation, you only need the last candle EMA and the current candle Close price.  But this begs the chicken vs egg question, which is why I brought up initialization.  If it helps, you can review my [manual calculation of EMA in Excel](https://github.com/DaveSkender/Stock.Indicators/raw/ef3d4b2f641b0742ba19f5a22514f932c63b6722/indicators/Ema/Ema.Calc.xlsx) (specifically the "ema" column).

**Reply by CyberNomadDev** [reply]
Thanks Dave
